### PR TITLE
AUTH-96: Add user info endpoint to CAS/AC 

### PR DESCRIPTION
**Pull request by bbooth**
Add user info endpoint to CAS/AC that allows someone to convert a JWT into user information.

Verification:
- Login into CAS using https://devcas.infusiontest.com:7443/login?service=http://localhost:8080/jwt
- This will redirect to nowhere but it will have a JWT in the `ticket` query param
- Make a call using postman using the JWT as a `Authorization: Bearer {token}` header
```
GET /jwt HTTP/1.1
Host: devcas.infusiontest.com:7443
Authorization: Bearer eyJhbGciOiJIUzUxMiJ9.WlhsS05tRllRV2xQYVVwRlVsVlphVXhEU21oaVIyTnBUMmxLYTJGWVNXbE1RMHBzWW0xTmFVOXBTa0pOVkVrMFVUQktSRXhWYUZSTmFsVXlTVzR3TGk0d1J6SnFOWGxQWlhKNmNEUlpjM04xWW01aVNGaFJMa1JTVkc5dFJVdGtiV2N4YUhadlJrdHpaR0pZWTBneloyOWZURU4xZFY5R2FXOTZNVVJ0Vmt4SmQzQmhTMDlOWDBzNFkzQXdOVmRHVjFjMWIwcHJXRkpvTmxkU1NrbzJOVkZzUVRWNFdHVkRORlp3ZWxsR09IQm1jMk53Y1RKQmFIbE1RV2h5Y1VWSExXMDVla0UzYmpsWk1rdzVVMVpQYVhkdlozQkdjQzFUVWxCSVVVUjNTMkYwV1ZwRWMwTjVOWGt5WjBSaWJ6azJlVTlOVFdORGVIVk9PRGRuZUhkRVNYRlpYek5oYmxoeFUyUldNemwyT1VkVFpGQkZTRmxLTUMxMlFWOVplR3c1ZGt0Q1RXYzBjakIzT0djeVRGZDZORkJsZUdkTmNtaFZXWE13WnpoMlkyVjBlVkZHVlRORU5YQlNVSHBIWlhscVgzUnJTRzVvV25obk5UWTRNMDlmV0RadWFrWlJSbWd3U2xOV1dWaGZkVzR6UWtWTFpGbHlVelZ4TjFoVGFVSjZURzg0TTBSR1JsVXpOR3c0UWtSMmVGbzFVV2M0WkhKSExVSjBZM3ByV0hOM1JERnlTMUUyVkRWMGFHUk5OeTFzZVZZeFkwUlVSMkZwYlVGVlgxbGFjVzVqVVRac1VrTk5jbVJaV1VneFVFeERaVzFJVG5keVJYVTNaRE5xYTJSSE9XTmxRUzFvZEhKSVFUaFdjbGhTWldOeFVsOUJaMUJNWlZJNExWVkthbVY1VEU5blRHWlVRVlUwVVZveE4wUXphVkF3V1RaRmRrTllhRWhOWjJwVU4wNWZSMngzVTJGSlRWaERXbkF4Umt4b1Vtc3pjbHA0VlVSc01WWjVORmhIWld4ZlprcHpTVUU1TW14cGRrVkhRM2w1Y1MxR1lrbFhTbFpTVWpScVZrdFhaM1JrTVhGdlRXWTFWa0ZHU2xGTVVqY3hjMUJ3YUROUVZFOTZhVmxDVkdNemMxZGlUbEEwWmpKcFpYcDZlV2xzTTJkVFdYaE1RWG96VDBwUFluVjBhVW8wUVVWaGFFRkxaSFZTY210bk1uaEpSR2hZY1VsbVJISlBhMEV0UmtzMlVVMDFUVmcyTTJVNE1XMVNWVmRVU1dVNVdHdEhUaTF1WTBVdFZtdFVVVVp6VUdVeFFpMTRhbUZsY2toblltWnNWbkpVWjE4eGNVY3ROVWR6VWxSWWFHZHZTWGhtVkdkd2NGSlllSE5ZVlRGdU1URmZYM1Z6UkdVek5rRkJkRVUzZGxRemRqTnpVVUpqV1hWRlJIRnlSVmRSTmtWYU1YcEhZVTl6UjBSUVJqSnFVSGszVms5MGRrOVFlRGRFU0VoRE0yOVpVRkZWTnpoNGNVeEtlbFEwV2pNNVVtSldiMmMwYlZWUVZEaFVSa0ZtWkZSV1NIVkpTM1p5WTFscWRWWjZaelo2TVRWdlNrSnhRMk5SYjBkSldFcFJSbVZhUjJ0WE1XWmlVMVZRZVdkVWFscEphSHBTUlZoMlpHVm5MbnAxWVU5cVFXazRlbDlKTTBwbk9UZzVhR3RHVTNjPQ.u_2Wdrw0Loytygz9gs2whUq-YJvZiOxKDRgwxYmeD2w7zhe8IAW5fSseWc8QvT1nFWPQW8ZltjpHgx_EHUJOrw
Cache-Control: no-cache
Postman-Token: 723788b5-116f-cf9b-4672-5b7ac06d6746
```

You should get a valid json response that looks like this:
```
{
    "sub": "21",
    "lastName": "Booth1",
    "isFromNewLogin": "true",
    "authenticationDate": "2017-08-01T21:46:49.555-07:00[America/Phoenix]",
    "displayName": "Bradley1 Booth1",
    "successfulAuthenticationHandlers": "Infusionsoft Authentication Handler",
    "iss": "https://devcas.infusiontest.com:7443",
    "authorities": [
        "ROLE_CAS_MARKETING_ADMIN",
        "ROLE_CAS_STEALTH_CRM",
        "ROLE_CAS_SUPPORT_TIER_2",
        "ROLE_CAS_LISTING_ADMIN_MARKETPLACE",
        "ROLE_CAS_ADMIN",
        "ROLE_CAS_USER",
        "ROLE_CAS_LISTING_PUBLISHER_MARKETPLACE",
        "ROLE_CAS_SUPPORT_TIER_1"
    ],
    "aud": "http://localhost:8080/jwt",
    "firstName": "Bradley1",
    "authenticationMethod": "Infusionsoft Authentication Handler",
    "longTermAuthenticationRequestTokenUsed": "false",
    "id": "21",
    "accounts": "[{\"appType\":\"CRM\",\"appName\":\"a46\",\"appUsername\":\"bradb@infusionsoft.com\",\"appAlias\":\"a46n\",\"appUrl\":\"https://a46.infusiontest.com:8443\"},{\"appType\":\"CRM\",\"appName\":\"a54\",\"appUsername\":\"bbooth\",\"appAlias\":\"a54\",\"appUrl\":\"https://a54.infusiontest.com:8443\"},{\"appType\":\"CRM\",\"appName\":\"a56\",\"appUsername\":\"bradb@infusionsoft.com\",\"appAlias\":\"a56\",\"appUrl\":\"https://a56.infusiontest.com:8443\"},{\"appType\":\"CRM\",\"appName\":\"devbox\",\"appUsername\":\"bradb@infusionsoft.com\",\"appAlias\":\"devbox\",\"appUrl\":\"https://devbox.infusiontest.com:8443\"},{\"appType\":\"CRM\",\"appName\":\"g36\",\"appUsername\":\"bradb@infusionsoft.com\",\"appAlias\":\"g36\",\"appUrl\":\"https://g36.infusiontest.com:8443\"},{\"appType\":\"CRM\",\"appName\":\"g46\",\"appUsername\":\"bradb@infusionsoft.com\",\"appAlias\":\"g46\",\"appUrl\":\"https://g46.infusiontest.com:8443\"},{\"appType\":\"CRM\",\"appName\":\"mashery1\",\"appUsername\":\"bradb@infusionsoft.com\",\"appAlias\":\"mashery1test\",\"appUrl\":\"https://mashery1.infusiontest.com:8443\"}]",
    "exp": 1501908475,
    "iat": 1501649275,
    "jti": "ST-1-HZP4VHvbVGxHdaRn7fmt-devcas.infusiontest.com",
    "email": "bradb@infusionsoft.com"
}
```

Validate bad data:
- No Authorization Header should return 401.
- Authorization Header does not start with `Bearer` should return 401.
- Authorization Header present but missing but token has invalid format should return 401 (remove everything in the token before the `.`)
- Valid format for JWT but token can't be decrypted should return 401 (add a `-bad` to the end of the token so it fails decryption).

### CODE REVIEW

**Comment by kevin-wise-infusionsoft** [comment]
Just had a thought. This is a public endpoint, so it could be used to decrypt any of our JWTs.  That seems to defeat the purpose of encrypting it to begin with.  Should we authenticate the calling service somehow?

**Comment by bbooth** [comment]
@kevin-wise-infusionsoft I don't think it is much different than a get user info endpoint that takes the jwt or an access token.  If the concern is decryption vs decryption and then use our own DTO, that maybe something to consider.

File: cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @ResponseBody
+    public ResponseEntity decrypt(HttpServletRequest request, HttpServletResponse response, Locale locale) throws CredentialsException, HttpAction {
+        J2EContext context = new J2EContext(request, response);
+
+        try {
+            HeaderExtractor headerExtractor = new HeaderExtractor("Authorization", "Bearer ", "(CAS) JWT Validator");
+            TokenCredentials tokenCredentials = headerExtractor.extract(context);
+
+            if (tokenCredentials != null) {
+                String decrypted = tokenTicketCipherExecutor.decode(tokenCredentials.getToken());
```
Comment (kevin-wise-infusionsoft) [inline cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java]: Does `tokenTicketCipherExecutor.decode` validate that the token is valid (not expired, etc)?  It probably should validate that it's not expired if we're essentially using it as credentials on this endpoint.

Comment (bbooth) [inline cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java]: I will check that right now.

File: cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    @RequestMapping(method = RequestMethod.GET, produces = "application/json")
+    @ResponseBody
+    public ResponseEntity decrypt(HttpServletRequest request, HttpServletResponse response, Locale locale) throws CredentialsException, HttpAction {
+        J2EContext context = new J2EContext(request, response);
+
+        try {
+            HeaderExtractor headerExtractor = new HeaderExtractor("Authorization", "Bearer ", "(CAS) JWT Validator");
+            TokenCredentials tokenCredentials = headerExtractor.extract(context);
+
+            if (tokenCredentials != null) {
```
Comment (MikeChristianson) [inline cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java]: I don't know if this is any better, but the nested `null` checks kind of bug me.

```java
        try {
            HeaderExtractor headerExtractor = new HeaderExtractor("Authorization", "Bearer ", "(CAS) JWT Validator");
            TokenCredentials token = extract(headerExtractor, context).orElseThrow(() -> new BadJWEException("Invalid JWE Token"));
            String decrypted = decode(tokenTicketCipherExecutor, token).orElseThrow(() -> new BadJWEException("Invalid JWE Token"));
            return new ResponseEntity<>(decrypted, HttpStatus.OK);
        } catch (CredentialsException | IllegalArgumentException | BadJWEException e) {
            return logAndReturnError(e, "cas.exception.jwt.invalid", new Object[]{e.getMessage()}, locale, HttpStatus.UNAUTHORIZED);
        }
```

```java
    private Optional<String> decode(TokenTicketCipherExecutor tokenTicketCipherExecutor, TokenCredentials nonNullToken) {
        return Optional.ofNullable(tokenTicketCipherExecutor.decode(nonNullToken));
    }

    private Optional<TokenCredentials> extract(HeaderExtractor headerExtractor, J2EContext context) throws HttpAction, CredentialsException {
        return Optional.ofNullable(headerExtractor.extract(context));
    }
```

File: cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        } catch (CredentialsException | IllegalArgumentException | BadJWEException e) {
+            return logAndReturnError(e, "cas.exception.jwt.invalid", new Object[]{e.getMessage()}, locale, HttpStatus.UNAUTHORIZED);
+        }
+
+    }
+
+    private ResponseEntity<APIErrorDTO> logAndReturnError(Exception e, String code, Object[] args, Locale locale, HttpStatus httpStatus) {
+        String errorMessage = messageSource.getMessage(code, args, locale);
+        log.error(errorMessage, e);
+        return new ResponseEntity<>(new APIErrorDTO(code, errorMessage), httpStatus);
```
Comment (MikeChristianson) [inline cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java]: Is there an alternative to this Deprecated DTO?

Comment (bbooth) [inline cas/src/main/java/org/apereo/cas/infusionsoft/web/controllers/JwtController.java]: The DTO probably should not be deprecated anymore.  Kevin went through and deprecated a bunch of stuff when we split CAS and AC.

**Comment by bbooth** [comment]
Updated to validate the JWT after decryption.

**Pull request merged**